File: typescript/src/E3.ts
// yarn run build && node build/src/E3.js

import { Rec, Ret, Delay, runRec } from './Trampoline';

type List<T> = Nil | Cons<T>;

type Nil = { tag: 'Nil' };
const Nil: List<never> = { tag: 'Nil' };
type Cons<T> = { tag: 'Cons', head: T, tail: List<T> };
const Cons = <T>(head: T, tail: List<T>): List<T> => ({ tag: 'Cons', head, tail });

const tail = <T>(l: List<T>): List<T> =>
  l.tag === 'Nil' ? Nil : l.tail;

const setHead = <T>(l: List<T>, h: T): List<T> =>
  l.tag === 'Nil' ? Nil : Cons(h, l.tail);

const drop = <T>(l: List<T>, n: number): List<T> =>
  l.tag === 'Nil' ? Nil : n == 0 ? l : drop(l.tail, n - 1);

const dropWhile = <T>(l: List<T>, f: (x: T) => boolean): List<T> =>
  l.tag === 'Nil' ? Nil : f(l.head) ? dropWhile(l.tail, f) : l;

const init = <T>(l: List<T>): List<T> =>
  l.tag === 'Nil' ? Nil : l.tail.tag === 'Nil' ? Nil : Cons(l.head, init(l.tail));

const foldRight = <A, B>(l: List<A>, x: B, f: (h: A, a: B) => B): B =>
  l.tag === 'Nil' ? x : f(l.head, foldRight(l.tail, x, f));

const length2 = <T>(l: List<T>): number => foldRight(l, 0, (_, b) => b + 1);

// use trampoline to make it tail recursive
const foldLeft = <A, B>(l: List<A>, x: B, f: (a: B, h: A) => B) => {
  const loop = (l: List<A>, a: B): Rec<B> =>
    l.tag === 'Nil' ? Ret(a) : Delay(() => loop(l.tail, f(a, l.head)));
  return runRec(loop(l, x));
};

const sum = (l: List<number>): number => foldLeft(l, 0, (a, b) => a + b);
const product = (l: List<number>): number => foldLeft(l, 1, (a, b) => a * b);
const length3 = <T>(l: List<T>): number => foldLeft(l, 0, (a, _) => a + 1)

File: scala/E3.scala
// scala E3.scala

sealed trait List[+A]
case object Nil extends List[Nothing]
case class Cons[+A](head: A, tail: List[A]) extends List[A]

object List {

  def tail[T](l: List[T]): List[T] =
    l match {
      case Nil => Nil
      case Cons(_, t) => t
    }

  def setHead[T](l: List[T], h: T): List[T] =
    l match {
      case Nil => Nil
      case Cons(_, t) => Cons(h, t)
    }

  @annotation.tailrec
  def drop[T](l: List[T], n: Int): List[T] =
    (l, n) match {
      case (Nil, _) => Nil
      case (l, 0) => l
      case (Cons(_, t), n) => drop(t, n - 1)
    }

  def dropWhile[T](l: List[T], f: T => Boolean): List[T] = {
    @annotation.tailrec
    def loop(l: List[T]): List[T] =
      l match {
        case Nil => Nil
        case Cons(h, t) => if (f(h)) loop(t) else l
      }

    loop(l)
  }

  // I was unable to write this tail-recursively nicely
  def init[T](l: List[T]): List[T] =
    l match {
      case Nil => Nil
      case Cons(_, Nil) => Nil
      case Cons(h, t) => Cons(h, init(t))
    }

}
